### DISCUSSION TITLE
Power waveform

### DISCUSSION THREAD

**Discussion by Blebowski** [parallaxsw/OpenSTA#68] [Q&A] [answered]
Hi,
does OpenSTA support dumping the power consumption waveform in some analog waveform format ?
If so, how do I do it ? If not, I dare to say its something worth adding.

**Comment by rovinski** [comment]
"analog waveform format" This just sounds like SPICE? Do you mean like per-cycle consumption instead of aggregated?

**Reply by Blebowski** [reply]
Hi, yes, I mean something like SPICE waveform where power consumption is displayed per each simulated time interval.

**Reply by rovinski** [reply, marked as answer]
OpenSTA and all similar digital simulation are cycle- or event-based tools. The best you will be able to do with them is a single-cycle simulation and get the power from that (which won't be very accurate). They are meant for longer simulations and an average power number.

For arbitrary timescales shorter than a clock cycle, you just need to go directly to SPICE. OpenSTA can at least help by using `write_path_spice` for subsets of the circuit you are interested in. Any circuit sufficiently large will take forever to simulate in SPICE anyways.

**Reply by Blebowski** [reply]
Thank you for answering.

A commercial flow I was using had option to specify "time interval" when doing power analysis in the STA tool.
I think, the tool then computed power in each time interval based on events / instances being activated at
the given time (based on VCD behavior).

Since STA is doing propagation of signals with delays, the tool "knows" that the dynamic power at particular
instance occurs (at the time when the change propagates through the instance). Thus, it can accumulate such
elementary dynamic power contributions and sum them at each time in waveform. The tool then dumped
a FSDB waveform, and we could see peak consumptions each time around the clock edge. The power 
consumption was spread a little around the clock edge due to clock skew. Obviously, this was not realistic
since it neglects impact of decoupling, but once we set the interval to average over whole clock cycle,
we got dynamic consumption waveform, that we were able to correlate quiet precisely to actual power-coarse
measurement at matching VT conditions.

**Reply by akashlevy** [reply]
I'd like to second the request for cycle-by-cycle (or windowed) power waveform, whenever resources permit. I may look into adding support for it myself depending on how much work it is

**Comment by jjcherry56** [comment]
OpenSTA could use the vcd file to generate a cycle by cycle power waveform. It is just an issue of resources to do it.
